src/Chapter3.hs

```diff
@@ -1004,8 +1106,79 @@ properties using typeclasses, but they are different data types in the end.
 
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
--}
 
+newtype Attack = Attack Int
+newtype Health = Health Int
+newtype Defense = Defense Int
+
+getValue (Attack v) = v
+getValue (Health v) = v
+getValue (Defense v) = v
+
+data KnightAction = KnightAttack | KnightDrinkPotion | KnightCastSpell
+data MonsterAction = MonsterAttack | MonsterRunAway
+
+data Knight = Knight
+    { knightHealth :: Health
+    , knightAttack :: Attack
+    , knightDefense :: Defense
+    }
+
+data Monster = Monster
+    { monsterHealth :: Health
+    , monsterAttack :: Attack
+    }
+
+class AttackClass a where
+    attack :: a -> Attack
+
+instance AttackClass Knight where
+    attack = knightAttack
+
+instance AttackClass Monster where
+    attack = monsterAttack
+
+class HealthClass a where
+    health :: a -> Health
+    updateHealth :: Health -> a -> a
+
+instance HealthClass Knight where
+    health = knightHealth
+    updateHealth h a = a { knightHealth = h }
+
+instance HealthClass Monster where
+    health = monsterHealth
+    updateHealth h a = a { monsterHealth = h }
+
+class DefenseClass a where
+    defense :: a -> Defense
+    updateDefense :: Defense -> a -> a
+
+instance DefenseClass Knight where
+    defense = knightDefense
+    updateDefense d a = a { knightDefense = d }
+
+instance DefenseClass Monster where
+    defense = 0 
+    updateDefense _ a = a
+
+castSpell a = Defense (1 + (getValue (defense a)))
+drinkPotion a = Health (1 + (getValue (health a)))
+attackOppenent f s = max 0 (getValue (health b) - (max 0 (getValue (attack f))- (getValue (defense s))))
```

Comment:
In order to master typeclasses, it is better to specify type signatures in here 🙂